packages/freezed/lib/src/freezed_generator.dart

```diff
@@ -248,26 +240,48 @@ Read here: https://github.com/rrousselGit/freezed/blob/master/packages/freezed/C
   }
 
   List<Parameter> _commonParametersBetweenAllConstructors(
-    List<ConstructorDetails> constructorsNeedsGeneration,
-  ) {
+    List<ConstructorDetails> constructorsNeedsGeneration, {
+    required bool allowCommonSuperType,
+  }) {
     return constructorsNeedsGeneration.first.parameters.allParameters
         .map((parameter) {
-          var hasAnyFinalProperty = false;
-          for (final constructor in constructorsNeedsGeneration) {
-            final matchingParameter =
-                constructor.parameters.allParameters.firstWhereOrNull((p) {
-              return p.name == parameter.name && p.type == parameter.type;
-            });
+          var anyMatchingPropertyIsFinal = false;
+          var isCommonWithDifferentNullability = false;
+
+          DartType? commonType;
+
+          for (final constructor in constructorsNeedsGeneration.skip(1)) {
+            final matchingParameter = constructor.parameters.allParameters
+                .firstWhereOrNull((p) => p.name == parameter.name);
 
             if (matchingParameter == null) return null;
-            if (matchingParameter.isFinal) hasAnyFinalProperty = true;
-          }
 
-          if (hasAnyFinalProperty) {
-            return parameter.copyWith(isFinal: true);
+            final commonTypeWithMatch =
+                parameter.parameterElement!.library!.typeSystem.leastUpperBound(
+              parameter.parameterElement!.type,
```

Comment:
Should this be commonType? (Initialize commonType with parameter.parameterElement!.type). That way it can handle >2 constructors with a common base type

Reply:
You could then remove the if (commonType != commonTypeWithMatch) check below.

Reply:
We would keep getting the supertype with each previous supertype until we hit Object. And if Object isn't the initial type we'll return since as per my other statement I think it's not desirable to have each common parameter as Object in the base mixin.